### PR TITLE
#3 Fetch m2m token

### PR DESCRIPTION
**Pull request by kaklakariada**
Part of #3

### CODE REVIEW

File: src/main/lua/exasol/adapter/databricks/authentication.lua [hunk 1]
```diff
@@ -0,0 +1,70 @@
+local http_client = require("exasol.adapter.databricks.http_client")
+local util = require("exasol.adapter.databricks.util")
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/authentication.lua:2]: We talked about this, please split and name appropriately.

File: src/main/lua/exasol/adapter/databricks/authentication.lua [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+local log = require("remotelog")
+
+local M = {}
+
+---@alias TokenProvider fun(): string
+
+---@param connection_details DatabricksConnectionDetails
+---@return string
+local function fetch_oauth_token(connection_details)
+    log.info("Fetching new OAuth M2M token")
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/authentication.lua:14]:
```suggestion
    log.trace("Fetching new OAuth M2M token")
```

File: src/main/lua/exasol/adapter/databricks/authentication.lua [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+---@param connection_details DatabricksConnectionDetails
+---@return TokenProvider
+local function create_m2m_token_provider(connection_details)
+    local token = nil
+    return function()
+        if token == nil then
+            token = fetch_oauth_token(connection_details)
+        else
+            log.debug("Token already available, no need to fetch it again")
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/authentication.lua:40]:
```suggestion
            log.trace("Token already available, no need to fetch it again")
```

File: src/main/lua/exasol/adapter/databricks/authentication.lua [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        end
+        return token
+    end
+end
+
+---@param token string
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/authentication.lua:46]:
```suggestion
---@param token string bearer token
```

File: src/main/lua/exasol/adapter/databricks/authentication.lua [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+---@return TokenProvider token_provider
+function M.create_token_provider(connection_details)
+    if connection_details.token then
+        return create_bearer_token_provider(connection_details.token)
+    end
+    if connection_details.oauth_client_id and connection_details.oauth_client_secret then
+        return create_m2m_token_provider(connection_details)
+    end
+    local exa_error = tostring(ExaError:new("E-VSDAB-20", "No Databricks credentials found."):add_mitigations(
+            "Specify token or OAuth M2 credentials as specified in the user guide."))
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/authentication.lua:65]:
```suggestion
            "Please provide token or OAuth M2 credentials as specified in the user guide."))
```

File: src/main/lua/exasol/adapter/databricks/http_client.lua [hunk 6]
```diff
@@ -3,6 +3,7 @@ local ExaError = require("ExaError")
 local http = require("socket.http")
 local socket = require("socket")
 local ssl = require("ssl")
+local ltn12 = require("ltn12")
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/http_client.lua:6]:
```suggestion
local ltn12 = require("ltn12")
```
Unused.

File: src/main/lua/exasol/adapter/databricks/http_client.lua [hunk 7]
```diff
@@ -92,6 +94,30 @@ function M._create_socket_factory(args)
     end
 end
 
+---@alias BodySource fun(): string?
+
+---Creates an ltn12 source for the given string data or `nil` if the data is nil.
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/http_client.lua:99]: Please add a link to the source documentation

File: src/main/lua/exasol/adapter/databricks/http_client.lua [hunk 8]
```diff
@@ -92,6 +94,30 @@ function M._create_socket_factory(args)
...
+---Based on https://github.com/lunarmodules/luasocket/blob/master/src/ltn12.lua#L118
+---@param data string?
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/http_client.lua:101]:
```suggestion
---@param data string? body data content
```

File: src/main/lua/exasol/adapter/databricks/http_client.lua [hunk 9]
```diff
@@ -92,6 +94,30 @@ function M._create_socket_factory(args)
...
+---@return BodySource? source
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/http_client.lua:102]:
```suggestion
---@return BodySource? source data iterator for body content blocks
```

File: src/main/lua/exasol/adapter/databricks/util.lua [hunk 10]
```diff
@@ -13,4 +13,31 @@ function M.map(tbl, f)
     return t
 end
 
+local BASE64_CHARS<const> = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
+---Base64 encode a given string.
+---Based on https://stackoverflow.com/a/35303321
+---@param data string
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/util.lua:19]:
```suggestion
---@param data string original data
```

File: src/main/lua/exasol/adapter/databricks/util.lua [hunk 11]
```diff
@@ -13,4 +13,31 @@ function M.map(tbl, f)
...
+---@return string base64_encoded_data
```
Comment (redcatbear) [inline src/main/lua/exasol/adapter/databricks/util.lua:20]:
```suggestion
---@return string base64_encoded_data data as Base64-encoded string
```

File: src/test/java/com/exasol/adapter/databricks/databricksfixture/DatabricksFixture.java [hunk 12]
```diff
@@ -86,12 +86,20 @@ private Connection getJdbcConnection() {
...
         final URI databricksUri = config.getDatabricksHostUri();
         final String hostName = databricksUri.getHost();
         final int port = databricksUri.getPort() < 0 ? 443 : databricksUri.getPort();
         final String httpPath = "/sql/1.0/warehouses/" + getEndpoint().getId();
-        return String.format("jdbc:databricks://%s:%d/default;transportMode=http;ssl=1;AuthMech=3;httpPath=%s;",
-                hostName, port, httpPath);
+        final String oauthCredentials = authMode == AuthMode.OAUTH_M2M
+                ? "OAuth2ClientId=%s;OAuth2Secret=%s;".formatted(config.getDatabricksOauthClientId(),
+                        config.getDatabricksOauthSecret())
+                : "";
```
Comment (redcatbear) [inline src/test/java/com/exasol/adapter/databricks/databricksfixture/DatabricksFixture.java:100]:
```suggestion
        final String oauthCredentials = (authMode == AuthMode.OAUTH_M2M)
                ? "OAuth2ClientId=%s;OAuth2Secret=%s;".formatted(config.getDatabricksOauthClientId(),
                        config.getDatabricksOauthSecret())
                : "";
```

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: doc/user_guide/user_guide.md [hunk 13]
```diff
@@ -117,7 +117,7 @@ Create a named connection:
 
 ```sql
 CREATE OR REPLACE CONNECTION DATABRICKS_JDBC_CONNECTION
-TO 'jdbc:databricks://$WORKSPACE_HOST_NAME:$PORT/default;transportMode=http;ssl=1;AuthMech=3;httpPath=$HTTP_PATH;'
+TO 'jdbc:databricks://$WORKSPACE_HOST_NAME:$PORT/default;transportMode=http;ssl=1;AuthMech=$AUTH_MECH;Auth_Flow=$AUTH_FLOW;OAuth2ClientId=$OAUTH_CLIENT_ID;OAuth2Secret=$OAUTH_SECRET;httpPath=$HTTP_PATH;'
```
[the remark this answers is not in the document]
Comment (kaklakariada) [reply doc/user_guide/user_guide.md:120]: I created https://github.com/exasol/databricks-virtual-schema/issues/43 to address this.

**Review by redcatbear — APPROVED** [review]

**Pull request merged**